### ISSUE TITLE
Problems with umlauts

### ISSUE THREAD

**Issue by AdmiralPellaeon** [assimp/assimp#3340]
Hi,

when my file names cointains umlauts, e.g. an 'ä', then 'ReadFile' fails to load my files.
Therefore my first question: how does assimp expect the char*? Atm I use a "C" locale as current code page.

And my second question: in some issues here I read something of Unicode support of assimp. But I didn't find hints about this in the documentation. So how can I switch to unicode and what unicode version does assimp supprot (UTF8, UTF16?)

Best regards

**Comment by kimkulling** [comment]
Hi, yes I know about this issue. Which kind of model have you tried to load? I just want to make sure if this issue comes from the dedicated importer or from the global API.

Thanks!

**Comment by RichardTea** [comment]
Internally all strings passed as char* are UTF-8, however it it likely there are cases where the conversations are incorrect

**Comment by AdmiralPellaeon** [comment]
Hi,

sorry for the late reply. I was in holiday.

I load DirectX file format (*.X).

I looked into the assimp code and came to the same conclusion as RichardTea: Assimp expects UTF8 strings. When loading files a WinAPI function is used to convert UTF8 to widechar. 

I tried to convert my ANSI string to UTF8 directly and then to pass it to assimp, but I got always a wrong conversion. I suppose some problem with wrong code pages in my conversion, but I didn't figured the problem out. Now, I convert my ANSI to widestring. After this I convert the wide string to UTF8 with boost and this resulting string I forward to assimp. A little bit annoying, but it works.

**Comment by RichardTea** [comment]
It sounds like your real issue is that you're using ANSI WinAPIs to get your filename.
Don't do that unless you absolutely have to.
ANSI forces you to use codepages, which are a whole world of pain and literally impossible to fully get right.

The ANSI versions of the WinAPI cannot properly support conversions to Unicode (eg UTF-8) because Unicode can contain code points from multiple code pages, which is impossible in ANSI.

Use the Unicode WinAPI  which is UTF-16 (little endian), and then you can use WideCharToMultiByte(CP_UTF8, ...) for your conversion. (Note the explicit UTF-8!)

I have noticed that some versions of Visual Studio default your application to ANSI when creating a new project.
That may be your problem, so make sure it's set to Unicode.

**Comment by RichardTea** [comment]
btw - such an API change would not spare any conversions. Assimp would have to convert it to UTF-8 immediately anyway, because everything inside Assimp requires strings to be UTF-8

**Comment by AdmiralPellaeon** [comment]
> 
> 
> It sounds like your real issue is that you're using ANSI WinAPIs to get your filename.
> Don't do that unless you absolutely have to.
> ANSI forces you to use codepages, which are a whole world of pain and literally impossible to fully get right.
> 
> The ANSI versions of the WinAPI cannot properly support conversions from Unicode (eg UTF-8) because Unicode can contain code points from multiple code pages, which is impossible in ANSI.
> 
> Use the Unicode WinAPI which is UTF-16 (little endian), and then you can use WideCharToMultiByte(CP_UTF8, ...) for your conversion. (Note the explicit UTF-8!)
> 


Yes I know. But the relevant project isn't the newest one^^ Upgrading it to UTF16 is on the wish list :) And with Assimp 3 (and I think also Assimp 4) there were no problems with umlauts. Anyway, with the described workarounds it runs properly.

Regarding the API: Here I must disagree. At least under Windows, assimp (at least the ReadFile method) requests an UTF8 char, converts this UTF8 char to UTF16 and then opens the file. So under windows when using UTF16, which is the native encoding here, you have to convert your string to UTF8 when using assimp. And assimp is converting them back to UTF16. (the origin string format). So you could spare a conversion with an additional wchar API for ReadFile under Windows.

**Comment by tellypresence** [comment]
@AdmiralPellaeon Did the workarounds solve this for you?

**Comment by AdmiralPellaeon** [comment]
> [@AdmiralPellaeon](https://github.com/AdmiralPellaeon) Did the workarounds solve this for you?

A challenging question, the posts are 5 years old :D
But yes, as written, it worked fine with the workaround.

**Comment by tellypresence** [comment]
Closing, please reopen if it recurs
